skrub/_expressions/_evaluation.py

```diff
@@ -0,0 +1,711 @@
+import copy
+import functools
+import inspect
+import types
+import warnings
+from collections import defaultdict
+from types import SimpleNamespace
+
+from sklearn.base import BaseEstimator
+from sklearn.base import clone as skl_clone
+
+from . import _choosing
+from ._expressions import (
+    _BUILTIN_MAP,
+    _BUILTIN_SEQ,
+    Expr,
+    IfElse,
+    Var,
+    _Constants,
+)
+from ._utils import X_NAME, Y_NAME, simple_repr
+
+__all__ = [
+    "evaluate",
+    "clone",
+    "find_X",
+    "find_y",
+    "find_node_by_name",
+    "graph",
+    "nodes",
+    "clear_results",
+    "describe_steps",
+    "reachable",
+    "get_params",
+    "set_params",
+]
+
+
+def _as_gen(f):
+    if inspect.isgeneratorfunction(f):
+        return f
+
+    @functools.wraps(f)
+    def g(*args, **kwargs):
+        if False:
+            yield
+        return f(*args, **kwargs)
+
+    return g
+
+
+class _Computation:
+    def __init__(self, target, generator):
+        self.target_id = id(target)
+        self.generator = generator
+
+
+class CircularReferenceError(ValueError):
+    pass
+
+
+class _ExprTraversal:
```

Comment:
Overall, I think that this call needs a bit more comments. It is quite abstract code, and I fear that people reading it will get lost